README.rst
*********************
collective.prettydate
*********************

.. contents:: Table of Contents

Life, the Universe, and Everything
==================================

Fuzzy dates for Plone.

This package helps developers to solve some usability issues associated with the utilization of absolute date formating on sites used among different timezones.

When a date is printed as "01/02/2012" it could represent "February 1" or "January 2",
depending on the format used.
Also,
if the site is located on a different timezone,
it could take you to situations when today's date is printed as yesterday,
or tomorrow's date.

collective.prettydate represents a date on a relative format so it would be displayed as "4 hours ago",
"yesterday" or "last week",
which is easier to read and understand for most people.

This feature is really nice for news sites or sites with upcoming events.

Mostly Harmless
===============

.. image:: https://secure.travis-ci.org/collective/collective.prettydate.png?branch=master
    :alt: Travis CI badge
    :target: http://travis-ci.org/collective/collective.prettydate

.. image:: https://coveralls.io/repos/collective/collective.prettydate/badge.png?branch=master
    :alt: Coveralls badge
    :target: https://coveralls.io/r/collective/collective.prettydate

.. image:: https://pypip.in/d/collective.prettydate/badge.png
    :alt: Downloads
    :target: https://pypi.python.org/pypi/collective.prettydate/

Got an idea? Found a bug? Let us know by `opening a support ticket`_.

.. _`opening a support ticket`: https://github.com/collective/collective.prettydate/issues

Don't Panic
===========

Installation
------------

To enable this product in a buildout-based installation:

1. Edit your buildout.cfg and add ``collective.prettydate`` to the list of
   eggs to install ::

    [buildout]
    ...
    eggs =
        collective.prettydate

After updating the configuration you need to run ''bin/buildout'', which will
take care of updating your system.

Usage
-----

This product provides a utility which will convert a DateTime object into a human readable text::

    from zope.component import getUtility
    from collective.prettydate.interfaces import IPrettyDate
    date_utility = getUtility(IPrettyDate)

At this point, you can use the ``date`` method to convert the DateTime object::

    from DateTime import DateTime
    today = DateTime()
    str_date = date_utility.date(today)

In previous example, ``str_date`` will be "now".

The 'date' method also allows 2 additional parameters: 'short' and 'asdays'
which will modify the output to be in short format ('h' instead of 'hours',
'd' instead of 'days', etc) and whole days (it will use 'today' instead of any
amount of seconds, minutes or hours for current day). They both default to
"False".

Timezones
^^^^^^^^^

If your server timezone is configured correctly,
and the DateTime object contains valid timezone info,
then the text output will reflect the correct time.
You do not need to convert date and time between different timezones.

Output examples
^^^^^^^^^^^^^^^

* '4 hours ago'
* '4h ago' (short format)
* 'in 28 minutes'
* 'in 6 months'
* 'today'
* 'last week'
* 'yesterday'
* 'last year'

DateTime 3.0
^^^^^^^^^^^^

This package is fully compatible with `DateTime 3.0`_ which provides a significantly smaller memory footprint.

.. _`DateTime 3.0`: http://pypi.python.org/pypi/DateTime

Not entirely unlike
===================

`collective.js.timeago`_
    An alternative, and probably clever way of dealing with fuzzy dates in Plone.
    This package registers the `timeago`_ JQuery plugin into the resource registry so convertion takes place client-side.

.. _`collective.js.timeago`: https://pypi.python.org/pypi/collective.js.timeago
.. _`timeago`: http://timeago.yarp.com/


CHANGES.rst
Changelog
=========

There's a frood who really knows where his towel is.

1.2.3 (unreleased)
------------------

- Nothing changed yet.


1.2.2 (2014-11-12)
------------------

- Fixed German translation. [mbaechtold]


1.2.1 (2014-09-01)
------------------

- Add German translations. [jone]

- Fix package dependencies. [hvelarde]


1.2 (2013-03-19)
----------------

- Added Finnish translation. [petri]

- Added Italian translation. [gborelli]

- Tested compatibility with Plone 4.3. [hvelarde]

- Updated documentation. [hvelarde]

- Tested compatibility with Plone 4.0. [hvelarde]


1.1 (2012-05-29)
----------------

- Removed unused test module. [hvelarde]

- Remove unnecessary generic setup profile. [nueces]

- Package no longer depends on five.grok (fixes #3). [frapell]


1.0 (2012-05-04)
----------------

- Tested compatibility with Plone 4.2. [hvelarde]

- Tested compatibility with DateTime 3.0 which provides a significantly
  smaller memory footprint. [hvelarde]


1.0rc1 (2012-02-21)
-------------------

- Initial release.


src/collective/prettydate/tests/test_setup.py
# -*- coding: utf-8 -*-

import unittest2 as unittest

from plone.app.testing import TEST_USER_ID
from plone.app.testing import setRoles

from collective.prettydate.testing import INTEGRATION_TESTING


class InstallTest(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']


class UninstallTest(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        setRoles(self.portal, TEST_USER_ID, ['Manager'])


def test_suite():
    return unittest.defaultTestLoader.loadTestsFromName(__name__)


src/collective/prettydate/tests/test_utility.py
# -*- coding: utf-8 -*-

import unittest2 as unittest

from zope.component import getUtility
from collective.prettydate.interfaces import IPrettyDate

from collective.prettydate.testing import INTEGRATION_TESTING

from DateTime import DateTime


class UtilityTest(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        self.request = self.layer['request']
        self.utility = getUtility(IPrettyDate)

        self.second = (
            DateTime('2012/12/21 00:00:01') - DateTime('2012/12/21 00:00:00'))
        self.minute = (
            DateTime('2012/12/21 00:01:00') - DateTime('2012/12/21 00:00:00'))
        self.hour = (
            DateTime('2012/12/21 01:00:00') - DateTime('2012/12/21 00:00:00'))
        self.day = (
            DateTime('2012/12/21 00:00:00') - DateTime('2012/12/20 00:00:00'))
        self.week = self.day * 7
        self.month = self.day * 31
        self.year = self.day * 365

    def test_long_past_time(self):
        # Let's ask for a date up to 9 seconds ago, we should get "now"
        for i in range(0, 10):
            now = DateTime()
            result = self.utility.date(now - (self.second * i))
            self.assertEqual(self.portal.translate(result), u'now')

        # From 10 seconds and up to 1 minute, we should get the number of
        # seconds ago
        for i in range(10, 60):
            now = DateTime()
            result = self.utility.date(now - (self.second * i))
            self.assertEqual(self.portal.translate(result),
                             u'%s seconds ago' % i)

        # From 1 minute and up to 2 minutes, we get "a minute ago"
        now = DateTime()
        result = self.utility.date(now - (self.minute))
        self.assertEqual(self.portal.translate(result), u'a minute ago')
        result = self.utility.date(now - (self.minute * 2 - self.second))
        self.assertEqual(self.portal.translate(result), u'a minute ago')

        # From 3 minutes and up to an hour, we get the number of minutes ago
        for i in range(2, 60):
            now = DateTime()
            result = self.utility.date(now - (self.minute * i))
            self.assertEqual(self.portal.translate(result),
                             u'%s minutes ago' % i)

        # From 1 hour and up to 2 hours, we get "an hour ago"
        now = DateTime()
        result = self.utility.date(now - (self.hour))
        self.assertEqual(self.portal.translate(result), u'an hour ago')
        result = self.utility.date(now - (self.hour * 2 - self.second))
        self.assertEqual(self.portal.translate(result), u'an hour ago')

        # From 3 hours and up to a day, we get the number of hours ago
        for i in range(3, 24):
            now = DateTime()
            result = self.utility.date(now - (self.hour * i))
            self.assertEqual(self.portal.translate(result),
                             u'%s hours ago' % i)

    def test_long_past_date(self):
        # Let's ask for yesterday
        now = DateTime()
        result = self.utility.date(now - (self.day))
        self.assertEqual(self.portal.translate(result), u'yesterday')

        # And for the day before
        now = DateTime()
        result = self.utility.date(now - (self.day * 3 - self.second))
        self.assertEqual(self.portal.translate(result), u'day before')

        # Now from 3 to up to 6 days, we get how many days ago
        for i in range(3, 7):
            now = DateTime()
            result = self.utility.date(now - (self.day * i))
            self.assertEqual(self.portal.translate(result), u'%s days ago' % i)

        # Let's check for last week
        now = DateTime()
        result = self.utility.date(now - (self.week))
        self.assertEqual(self.portal.translate(result), u'last week')
        now = DateTime()
        result = self.utility.date(now - (self.week * 2 - self.second))
        self.assertEqual(self.portal.translate(result), u'last week')

        # From 2 and up to 4 weeks we get the number of weeks ago
        for i in range(2, 5):
            now = DateTime()
            result = self.utility.date(now - (self.week * i))
            self.assertEqual(self.portal.translate(result),
                             u'%s weeks ago' % i)

        # Let's check for last month
        now = DateTime()
        result = self.utility.date(now - (self.month))
        self.assertEqual(self.portal.translate(result), u'last month')
        now = DateTime()
        result = self.utility.date(now - (self.month * 2 - self.day * 2))
        self.assertEqual(self.portal.translate(result), u'last month')

        # From 2 and up to 12 months we get the number of months ago
        for i in range(2, 12):
            now = DateTime()
            result = self.utility.date(now - (self.month * i))
            self.assertEqual(self.portal.translate(result),
                             u'%s months ago' % i)

        # Let's check for last year
        now = DateTime()
        result = self.utility.date(now - (self.year))
        self.assertEqual(self.portal.translate(result), u'last year')
        now = DateTime()
        result = self.utility.date(now - (self.year * 2 - self.second))
        self.assertEqual(self.portal.translate(result), u'last year')

        # Over 2 years, we get the number of years ago, let's test it up to
        # 20 years
        for i in range(2, 20):
            now = DateTime()
            result = self.utility.date(now - (self.year * i))
            self.assertEqual(self.portal.translate(result),
                             u'%s years ago' % i)

    def test_long_future_time(self):
        # Let's ask for a date up to 10 seconds in the future, we should get
        # "now"
        for i in range(0, 11):
            now = DateTime()
            result = self.utility.date(now + (self.second * i))
            self.assertEqual(self.portal.translate(result), u'now')

        # From 12 seconds and up to 1 minute, we should get the number of
        # seconds it takes
        # XXX There's a slight mismatch in the calculus, and is a bit off for
        #     future seconds, so we are going to make a little adjustement
        #     and hope for the best :P
        for i in range(11, 60):
            now = DateTime()
            result = self.utility.date(now + (self.second * (i + 1)))
            self.assertEqual(self.portal.translate(result),
                             u'in %s seconds' % i)

        # From 1 minute and up to 2 minutes, we get "in a minute"
        now = DateTime()
        result = self.utility.date(now + (self.minute + self.second))
        self.assertEqual(self.portal.translate(result), u'in a minute')
        result = self.utility.date(now + (self.minute * 2 - self.second))
        self.assertEqual(self.portal.translate(result), u'in a minute')

        # From 3 minutes and up to an hour, we get the number of minutes to go
        for i in range(2, 60):
            now = DateTime()
            result = self.utility.date(now +
                                       ((self.minute + self.second / 2.0) * i))
            self.assertEqual(self.portal.translate(result),
                             u'in %s minutes' % i)

        # From 1 hour and up to 2 hours, we get "in an hour"
        now = DateTime()
        result = self.utility.date(now + (self.hour + self.second))
        self.assertEqual(self.portal.translate(result), u'in an hour')
        result = self.utility.date(now + (self.hour * 2 - self.second))
        self.assertEqual(self.portal.translate(result), u'in an hour')

        # From 3 hours and up to a day, we get the number of hours to go
        for i in range(3, 24):
            now = DateTime()
            result = self.utility.date(now + (self.hour * i))
            self.assertEqual(self.portal.translate(result), u'in %s hours' % i)

    def test_long_future_date(self):
        # Let's ask for tomorrow
        now = DateTime()
        result = self.utility.date(now + (self.day + self.second))
        self.assertEqual(self.portal.translate(result), u'tomorrow')

        # And for the day after
        now = DateTime()
        result = self.utility.date(now + (self.day * 3 - self.hour))
        self.assertEqual(self.portal.translate(result), u'day after')

        # Now from 3 to up to 6 days, we get how many days to go
        for i in range(3, 7):
            now = DateTime()
            result = self.utility.date(now + ((self.day + self.second) * i))
            self.assertEqual(self.portal.translate(result), u'in %s days' % i)

        # Let's check for next week
        now = DateTime()
        result = self.utility.date(now + (self.week + self.second))
        self.assertEqual(self.portal.translate(result), u'next week')
        now = DateTime()
        result = self.utility.date(now + (self.week * 2 - self.second))
        self.assertEqual(self.portal.translate(result), u'next week')

        # From 2 and up to 4 weeks we get the number of weeks to go
        for i in range(2, 5):
            now = DateTime()
            result = self.utility.date(now + ((self.week + self.second) * i))
            self.assertEqual(self.portal.translate(result), u'in %s weeks' % i)

        # Let's check for next month
        now = DateTime()
        result = self.utility.date(now + (self.month + self.second))
        self.assertEqual(self.portal.translate(result), u'next month')
        now = DateTime()
        result = self.utility.date(now + (self.month * 2 - self.day))
        self.assertEqual(self.portal.translate(result), u'next month')

        # From 2 and up to 12 months we get the number of months to go
        for i in range(2, 12):
            now = DateTime()
            result = self.utility.date(now + ((self.month + self.second) * i))
            self.assertEqual(self.portal.translate(result),
                             u'in %s months' % i)

        # Let's check for next year
        now = DateTime()
        result = self.utility.date(now + (self.year + self.second))
        self.assertEqual(self.portal.translate(result), u'next year')
        now = DateTime()
        result = self.utility.date(now + (self.year * 2 - self.second))
        self.assertEqual(self.portal.translate(result), u'next year')

        # Over 2 years, we get the number of years to go, let's test it up to
        # 20 years
        for i in range(2, 20):
            now = DateTime()
            result = self.utility.date(now + ((self.year + self.second) * i))
            self.assertEqual(self.portal.translate(result), u'in %s years' % i)

    def test_short_past_time(self):
        # Let's ask for a date up to 9 seconds ago, we should get "now"
        for i in range(0, 10):
            now = DateTime()
            result = self.utility.date(now - (self.second * i), short=True)
            self.assertEqual(self.portal.translate(result), u'now')

        # From 10 seconds and up to 1 minute, we should get the number of
        # seconds ago
        for i in range(10, 60):
            now = DateTime()
            result = self.utility.date(now - (self.second * i), short=True)
            self.assertEqual(self.portal.translate(result), u'%ss ago' % i)

        # From 1 minute and up to an hour, we get the number of minutes ago
        for i in range(1, 60):
            now = DateTime()
            result = self.utility.date(now - (self.minute * i), short=True)
            self.assertEqual(self.portal.translate(result), u'%sm ago' % i)

        # From 1 hour and up to a day, we get the number of hours ago
        for i in range(1, 24):
            now = DateTime()
            result = self.utility.date(now - (self.hour * i), short=True)
            self.assertEqual(self.portal.translate(result), u'%sh ago' % i)

    def test_short_past_date(self):
        # Let's ask for yesterday
        now = DateTime()
        result = self.utility.date(now - (self.day), short=True)
        self.assertEqual(self.portal.translate(result), u'yest')

        # And for the day before
        now = DateTime()
        result = self.utility.date(now - (self.day * 3 - self.second),
                                   short=True)
        self.assertEqual(self.portal.translate(result), u'2d ago')

        # Now from 3 to up to 6 days, we get how many days ago
        for i in range(3, 7):
            now = DateTime()
            result = self.utility.date(now - (self.day * i), short=True)
            self.assertEqual(self.portal.translate(result), u'%sd ago' % i)

        # From 1 and up to 4 weeks we get the number of weeks ago
        for i in range(1, 5):
            now = DateTime()
            result = self.utility.date(now - (self.week * i), short=True)
            self.assertEqual(self.portal.translate(result), u'%sw ago' % i)

        # From 1 and up to 12 months we get the number of months ago
        for i in range(1, 12):
            now = DateTime()
            result = self.utility.date(now - (self.month * i), short=True)
            self.assertEqual(self.portal.translate(result), u'%smo ago' % i)

        # Over 1 year, we get the number of years ago, let's test it up to
        # 20 years
        for i in range(1, 20):
            now = DateTime()
            result = self.utility.date(now - (self.year * i), short=True)
            self.assertEqual(self.portal.translate(result), u'%sy ago' % i)

    def test_short_future_time(self):
        # Let's ask for a date up to 10 seconds in the future, we should get
        # "now"
        for i in range(0, 11):
            now = DateTime()
            result = self.utility.date(now + (self.second * i), short=True)
            self.assertEqual(self.portal.translate(result), u'now')

        # From 12 seconds and up to 1 minute, we should get the number of
        # seconds it takes
        # XXX There's a slight mismatch in the calculus, and is a bit off for
        #     future seconds, so we are going to make a little adjustement
        #     and hope for the best :P
        for i in range(11, 60):
            now = DateTime()
            result = self.utility.date(now + (self.second * (i + 1)),
                                       short=True)
            self.assertEqual(self.portal.translate(result), u'in %ss' % i)

        # From 1 minute and up to an hour, we get the number of minutes to go
        for i in range(1, 60):
            now = DateTime()
            date = now + ((self.minute + self.second / 2.0) * i)
            result = self.utility.date(date, short=True)
            self.assertEqual(self.portal.translate(result), u'in %sm' % i)

        # From 1 hour and up to a day, we get the number of hours to go
        for i in range(1, 24):
            now = DateTime()
            result = self.utility.date(now + ((self.hour + self.second) * i),
                                       short=True)
            self.assertEqual(self.portal.translate(result), u'in %sh' % i)

    def test_short_future_date(self):
        # Let's ask for tomorrow
        now = DateTime()
        result = self.utility.date(now + (self.day + self.second), short=True)
        self.assertEqual(self.portal.translate(result), u'tom')

        # And for the day after
        now = DateTime()
        result = self.utility.date(now + (self.day * 3 - self.hour),
                                   short=True)
        self.assertEqual(self.portal.translate(result), u'in 2d')

        # Now from 3 to up to 6 days, we get how many days to go
        for i in range(3, 7):
            now = DateTime()
            result = self.utility.date(now + ((self.day + self.second) * i),
                                       short=True)
            self.assertEqual(self.portal.translate(result), u'in %sd' % i)

        # From 1 and up to 4 weeks we get the number of weeks to go
        for i in range(1, 5):
            now = DateTime()
            result = self.utility.date(now + ((self.week + self.second) * i),
                                       short=True)
            self.assertEqual(self.portal.translate(result), u'in %sw' % i)

        # From 1 and up to 12 months we get the number of months to go
        for i in range(1, 12):
            now = DateTime()
            result = self.utility.date(now + ((self.month + self.second) * i),
                                       short=True)
            self.assertEqual(self.portal.translate(result), u'in %smo' % i)

        # Over 1 years, we get the number of years to go, let's test it up to
        # 20 years
        for i in range(1, 20):
            now = DateTime()
            result = self.utility.date(now + ((self.year + self.second) * i),
                                       short=True)
            self.assertEqual(self.portal.translate(result), u'in %sy' % i)

    def test_asdays(self):
        # For less than a day, it will always return today, and not hours.
        # for more than one day, it just behaves as the regular date. which
        # is already tested

        # let's repeat the time test, but always get "today"
        for i in range(0, 60):
            now = DateTime()
            result = self.utility.date(now - (self.second * i), asdays=True)
            self.assertEqual(self.portal.translate(result), u'today')

        for i in range(1, 60):
            now = DateTime()
            result = self.utility.date(now - (self.minute * i), asdays=True)
            self.assertEqual(self.portal.translate(result), u'today')

        for i in range(1, 24):
            now = DateTime()
            result = self.utility.date(now - (self.hour * i), asdays=True)
            self.assertEqual(self.portal.translate(result), u'today')

        # The same goes for the future
        for i in range(0, 60):
            now = DateTime()
            result = self.utility.date(now + (self.second * i), asdays=True)
            self.assertEqual(self.portal.translate(result), u'today')

        for i in range(1, 60):
            now = DateTime()
            result = self.utility.date(now + (self.minute * i), asdays=True)
            self.assertEqual(self.portal.translate(result), u'today')

        for i in range(1, 24):
            now = DateTime()
            result = self.utility.date(now + (self.hour * i), asdays=True)
            self.assertEqual(self.portal.translate(result), u'today')


src/collective/prettydate/utility.py
# -*- coding: utf-8 -*-

from DateTime import DateTime
from datetime import timedelta

from zope.interface import implements

from collective.prettydate.interfaces import IPrettyDate
from collective.prettydate import _


class PrettyDate(object):
    implements(IPrettyDate)

    def normalize(self, seconds, denominator=1):
        return unicode((seconds + denominator / 2) / denominator)

    def date(self, date, short=False, asdays=False):
        now = DateTime()
        time = DateTime(date)

        if time > now:
            past = False
            diff = timedelta(time - now)

        else:
            past = True
            diff = timedelta(now - time)

        seconds = diff.seconds
        days = diff.days

        result = u''

        if short:
            if days == 0 and not asdays:
                if seconds < 10:
                    result = _(u'now')
                elif seconds < 60:
                    if past:
                        result = _(
                            u"result_time_short_seconds_past",
                            default=u"${time}s ago",
                            mapping={u"time": self.normalize(seconds, 1)})
                    else:
                        result = _(
                            u"result_time_short_seconds_future",
                            default=u"in ${time}s",
                            mapping={u"time": self.normalize(seconds, 1)})

                elif seconds < 3600:
                    if past:
                        result = _(
                            u"result_time_short_minutes_past",
                            default=u"${time}m ago",
                            mapping={u"time": self.normalize(seconds, 60)})
                    else:
                        result = _(
                            u"result_time_short_minutes_future",
                            default=u"in ${time}m",
                            mapping={u"time": self.normalize(seconds, 60)})
                else:
                    if past:
                        result = _(
                            u"result_time_short_hours_past",
                            default=u"${time}h ago",
                            mapping={u"time": self.normalize(seconds, 3600)})
                    else:
                        result = _(
                            u"result_time_short_hours_future",
                            default=u"in ${time}h",
                            mapping={u"time": self.normalize(seconds, 3600)})
            else:
                if days == 0:
                    result = _(u'today')
                elif days == 1:
                    result = past and _(u'yest') or _(u'tom')
                elif days < 7:
                    if past:
                        result = _(
                            u"result_date_short_days_past",
                            default=u"${time}d ago",
                            mapping={u"time": self.normalize(days, 1)})
                    else:
                        result = _(
                            u"result_date_short_days_future",
                            default=u"in ${time}d",
                            mapping={u"time": self.normalize(days, 1)})
                elif days < 31:
                    if past:
                        result = _(
                            u"result_date_short_weeks_past",
                            default=u"${time}w ago",
                            mapping={u"time": self.normalize(days, 7)})
                    else:
                        result = _(
                            u"result_date_short_weeks_future",
                            default=u"in ${time}w",
                            mapping={u"time": self.normalize(days, 7)})
                elif days < 365:
                    if past:
                        result = _(
                            u"result_date_short_months_past",
                            default=u"${time}mo ago",
                            mapping={u"time": self.normalize(days, 30)})
                    else:
                        result = _(
                            u"result_date_short_months_future",
                            default=u"in ${time}mo",
                            mapping={u"time": self.normalize(days, 30)})
                else:
                    if past:
                        result = _(
                            u"result_date_short_years_past",
                            default=u"${time}y ago",
                            mapping={u"time": self.normalize(days, 365)})
                    else:
                        result = _(
                            u"result_date_short_years_future",
                            default=u"in ${time}y",
                            mapping={u"time": self.normalize(days, 365)})
        else:
            if days == 0 and not asdays:
                if seconds < 10:
                    result = _(u'now')
                elif seconds < 60:
                    if past:
                        result = _(
                            u"result_time_long_seconds_past",
                            default=u"${time} seconds ago",
                            mapping={u"time": self.normalize(seconds, 1)})
                    else:
                        result = _(
                            u"result_time_long_seconds_future",
                            default=u"in ${time} seconds",
                            mapping={u"time": self.normalize(seconds, 1)})
                elif seconds < 120:
                    result = past and _(u'a minute ago') or _(u'in a minute')
                elif seconds < 3600:
                    if past:
                        result = _(
                            u"result_time_long_minutes_past",
                            default=u"${time} minutes ago",
                            mapping={u"time": self.normalize(seconds, 60)})
                    else:
                        result = _(
                            u"result_time_long_minutes_future",
                            default=u"in ${time} minutes",
                            mapping={u"time": self.normalize(seconds, 60)})
                elif seconds < 7200:
                    result = past and _(u'an hour ago') or _(u'in an hour')
                else:
                    if past:
                        result = _(
                            u"result_time_long_hours_past",
                            default=u"${time} hours ago",
                            mapping={u"time": self.normalize(seconds, 3600)})
                    else:
                        result = _(
                            u"result_time_long_hours_future",
                            default=u"in ${time} hours",
                            mapping={u"time": self.normalize(seconds, 3600)})
            else:
                if days == 0:
                    result = _(u'today')
                elif days == 1:
                    result = past and _(u'yesterday') or _(u'tomorrow')
                elif days == 2:
                    result = past and _(u'day before') or _(u'day after')
                elif days < 7:
                    if past:
                        result = _(
                            u"result_date_long_days_past",
                            default=u"${time} days ago",
                            mapping={u"time": self.normalize(days, 1)})
                    else:
                        result = _(
                            u"result_date_long_days_future",
                            default=u"in ${time} days",
                            mapping={u"time": self.normalize(days, 1)})
                elif days < 14:
                    result = past and _(u'last week') or _(u'next week')
                elif days < 31:
                    if past:
                        result = _(
                            u"result_date_long_weeks_past",
                            default=u"${time} weeks ago",
                            mapping={u"time": self.normalize(days, 7)})
                    else:
                        result = _(
                            u"result_date_long_weeks_future",
                            default=u"in ${time} weeks",
                            mapping={u"time": self.normalize(days, 7)})
                elif days < 61:
                    result = past and _(u'last month') or _(u'next month')
                elif days < 365:
                    if past:
                        result = _(
                            u"result_date_long_months_past",
                            default=u"${time} months ago",
                            mapping={u"time": self.normalize(days, 30)})
                    else:
                        result = _(
                            u"result_date_long_months_future",
                            default=u"in ${time} months",
                            mapping={u"time": self.normalize(days, 30)})
                elif days < 730:
                    result = past and _(u'last year') or _(u'next year')
                else:
                    if past:
                        result = _(
                            u"result_date_long_years_past",
                            default=u"${time} years ago",
                            mapping={u"time": self.normalize(days, 365)})
                    else:
                        result = _(
                            u"result_date_long_years_future",
                            default=u"in ${time} years",
                            mapping={u"time": self.normalize(days, 365)})

        return result
